## BayWa r.e. Solar Trade - Subsidiary Performance Analysis

### PROJECT DESCRIPTION:
This analysis evaluates the financial and operational performance of three BayWa r.e.. Solar Trade subsidiaries (A, G, J) to provide strategic insights for C-Level executives and actionable recommendations for Operations managers.

### BUSINESS CONTEXT:
BayWa r.e. is a leading renewable energy company operating solar trade subsidiaries 
across different markets. Management needs to understand which subsidiaries are 
performing well, identify operational inefficiencies, and make informed investment 
decisions for future growth.

### ANALYTICAL APPROACH:
This analysis demonstrates real-world data analyst skills through:
- Comprehensive data quality assessment and cleaning
- Iterative investigation when initial results seemed suspicious  
- Critical thinking to uncover hidden operational issues
- Business segment analysis (retail vs. commercial solar projects)
- Product category performance evaluation
- Executive-level insights with actionable recommendations

KEY DATASETS:
- **Sales Orders**: Revenue transactions and order fulfillment data
- **Purchase Orders**: Procurement costs and supplier relationships  
- **Inventory**: Stock levels and cost management
- **Items**: Product catalog with categories and specifications

### DELIVERABLES:
1. Financial performance ranking of all subsidiaries
2. Operational efficiency analysis and improvement recommendations
3. Business model insights (retail vs. large commercial projects)
4. Interactive dashboard requirements for ongoing monitoring
5. Strategic recommendations for executive decision-making

In [1]:
# SETUP AND IMPORT LIBRARIES
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

### DATA LOADING

In [3]:
# Load Subsidiary A files
print("Loading Subsidiary A files:")
a_sales = pd.read_csv('A_sales_orders.csv')
a_purchase = pd.read_csv('A_purchase_orders.csv')
a_inventory = pd.read_csv('A_inventory.csv')
a_items = pd.read_csv('A_items.csv')
print(f"  A sales: {len(a_sales):,} rows")
print(f"  A purchases: {len(a_purchase):,} rows")
print(f"  A inventory: {len(a_inventory):,} rows")
print(f"  A items: {len(a_items):,} rows")

# Load Subsidiary G files
print("\nLoading Subsidiary G files:")
g_sales = pd.read_csv('G_sales_orders.csv')
g_purchase = pd.read_csv('G_purchase_orders.csv')
g_inventory = pd.read_csv('G_inventory.csv')
g_items = pd.read_csv('G_items.csv')
print(f"  G sales: {len(g_sales):,} rows")
print(f"  G purchases: {len(g_purchase):,} rows")
print(f"  G inventory: {len(g_inventory):,} rows")
print(f"  G items: {len(g_items):,} rows")

# Load Subsidiary J files
print("\nLoading Subsidiary J files:")
j_sales = pd.read_csv('J_sales_orders.csv')
j_purchase = pd.read_csv('J_purchase_orders.csv')
j_inventory = pd.read_csv('J_inventory.csv')
j_items = pd.read_csv('J_items.csv')
print(f"  J sales: {len(j_sales):,} rows")
print(f"  J purchases: {len(j_purchase):,} rows")
print(f"  J inventory: {len(j_inventory):,} rows")
print(f"  J items: {len(j_items):,} rows")

Loading Subsidiary A files:
  A sales: 32,844 rows
  A purchases: 2,171 rows
  A inventory: 397 rows
  A items: 636 rows

Loading Subsidiary G files:
  G sales: 135,546 rows
  G purchases: 14,244 rows
  G inventory: 3,069 rows
  G items: 1,737 rows

Loading Subsidiary J files:
  J sales: 143,161 rows
  J purchases: 9,512 rows
  J inventory: 2,743 rows
  J items: 1,707 rows


In [5]:
print("Subsidiary A:")
print(f"  Sales orders: {len(a_sales):,}")
print(f"  Purchase orders: {len(a_purchase):,}")
print(f"  Inventory items: {len(a_inventory):,}")
print(f"  Product catalog: {len(a_items):,}")

print("\nSubsidiary G:")
print(f"  Sales orders: {len(g_sales):,}")
print(f"  Purchase orders: {len(g_purchase):,}")
print(f"  Inventory items: {len(g_inventory):,}")
print(f"  Product catalog: {len(g_items):,}")

print("\nSubsidiary J:")
print(f"  Sales orders: {len(j_sales):,}")
print(f"  Purchase orders: {len(j_purchase):,}")
print(f"  Inventory items: {len(j_inventory):,}")
print(f"  Product catalog: {len(j_items):,}")

Subsidiary A:
  Sales orders: 32,844
  Purchase orders: 2,171
  Inventory items: 397
  Product catalog: 636

Subsidiary G:
  Sales orders: 135,546
  Purchase orders: 14,244
  Inventory items: 3,069
  Product catalog: 1,737

Subsidiary J:
  Sales orders: 143,161
  Purchase orders: 9,512
  Inventory items: 2,743
  Product catalog: 1,707


In [6]:
# Show sample columns
print("\nSample of sales data columns:")
print("A_sales columns:", list(a_sales.columns))


Sample of sales data columns:
A_sales columns: ['guid', 'headerGuid', 'itemGuid', 'locationGuid', 'inventoryGuid', 'tranID', 'linePosition', 'invoiceType', 'status', 'qtyOrdered', 'qtyOpen', 'qtyReceived', 'qtyBilled', 'amount', 'itemRate', 'createdDate', 'tranDate', 'actualShipDate', 'shipDate', 'expectedReceiptDate', 'is_web_order', 'is_intercompany', 'SubsidiaryID']


### BASIC DATA QUALITY CHECKS

In [7]:
# Check missing values in key columns
print("MISSING VALUES CHECK:")
print("\nSales data - 'amount' column:")
print(f"  A_sales missing: {a_sales['amount'].isnull().sum()}")
print(f"  G_sales missing: {g_sales['amount'].isnull().sum()}")
print(f"  J_sales missing: {j_sales['amount'].isnull().sum()}")

print("\nPurchase data - 'amount' column:")
print(f"  A_purchase missing: {a_purchase['amount'].isnull().sum()}")
print(f"  G_purchase missing: {g_purchase['amount'].isnull().sum()}")
print(f"  J_purchase missing: {j_purchase['amount'].isnull().sum()}")

MISSING VALUES CHECK:

Sales data - 'amount' column:
  A_sales missing: 0
  G_sales missing: 0
  J_sales missing: 0

Purchase data - 'amount' column:
  A_purchase missing: 0
  G_purchase missing: 0
  J_purchase missing: 0


In [8]:
# Check for unusual amounts (zero/negative)
print("\nUNUSUAL VALUES CHECK:")
print("Sales amounts = 0:")
print(f"  A: {(a_sales['amount'] == 0).sum()}")
print(f"  G: {(g_sales['amount'] == 0).sum()}")
print(f"  J: {(j_sales['amount'] == 0).sum()}")

print("\nPurchase amounts = 0:")
print(f"  A: {(a_purchase['amount'] == 0).sum()}")
print(f"  G: {(g_purchase['amount'] == 0).sum()}")
print(f"  J: {(j_purchase['amount'] == 0).sum()}")


UNUSUAL VALUES CHECK:
Sales amounts = 0:
  A: 7445
  G: 1969
  J: 3406

Purchase amounts = 0:
  A: 670
  G: 981
  J: 408


Key Data Quality Issues:
- Zero amounts: All subsidiaries have sales orders with €0 amounts (A has 7,445!)
- Date ranges: Data covers roughly 2022-2023 period
- Zero purchases: Some purchase orders also have €0 amounts

In [9]:
# Check date ranges
print("\nDATE RANGES:")
print(f"A_sales: {a_sales['tranDate'].min()} to {a_sales['tranDate'].max()}")
print(f"G_sales: {g_sales['tranDate'].min()} to {g_sales['tranDate'].max()}")
print(f"J_sales: {j_sales['tranDate'].min()} to {j_sales['tranDate'].max()}")


DATE RANGES:
A_sales: 2022-10-05 to 2023-12-31
G_sales: 2022-01-03 to 2023-12-29
J_sales: 2022-01-01 to 2023-12-31


### Checking Statistics

In [10]:
# Sales amount statistics
print("SALES AMOUNT STATISTICS:")
print("\nSubsidiary A Sales:")
print(a_sales['amount'].describe())

print("\nSubsidiary G Sales:")
print(g_sales['amount'].describe())

print("\nSubsidiary J Sales:")
print(j_sales['amount'].describe())

SALES AMOUNT STATISTICS:

Subsidiary A Sales:
count      32844.00
mean       13154.33
std       320551.51
min        -7110.78
25%           42.39
50%          847.80
75%         4713.14
max     48243171.48
Name: amount, dtype: float64

Subsidiary G Sales:
count    135546.00
mean       9904.76
std       55259.44
min       -6381.33
25%          76.87
50%         509.94
75%        3175.76
max     5677990.16
Name: amount, dtype: float64

Subsidiary J Sales:
count     143161.00
mean       11975.94
std        84608.88
min     -1059675.86
25%          348.41
50%         1816.14
75%         5487.02
max     11457820.31
Name: amount, dtype: float64


In [11]:
# Purchase amount statistics
print("\nPURCHASE AMOUNT STATISTICS:")
print("\nSubsidiary A Purchases:")
print(a_purchase['amount'].describe())

print("\nSubsidiary G Purchases:")
print(g_purchase['amount'].describe())

print("\nSubsidiary J Purchases:")
print(j_purchase['amount'].describe())


PURCHASE AMOUNT STATISTICS:

Subsidiary A Purchases:
count        2171.00
mean       313464.26
std       3811995.05
min             0.00
25%             0.00
50%          9647.59
75%         97214.40
max     131516728.63
Name: amount, dtype: float64

Subsidiary G Purchases:
count     14244.00
mean      79980.35
std      191334.59
min           0.00
25%        1224.60
50%        8809.77
75%       67239.96
max     5847896.06
Name: amount, dtype: float64

Subsidiary J Purchases:
count      9512.00
mean     148511.52
std      271429.26
min           0.00
25%        3536.27
50%       32117.49
75%      256646.02
max     7635585.10
Name: amount, dtype: float64


### Checking for Outliers:

In [12]:
# Check for extreme outliers in sales data
print("SALES OUTLIER ANALYSIS:")

for name, data in [('A', a_sales), ('G', g_sales), ('J', j_sales)]:
    amounts = data['amount']
    
    # Calculate quartiles
    q1 = amounts.quantile(0.25)
    q3 = amounts.quantile(0.75)
    iqr = q3 - q1
    
    # Calculate outlier boundaries
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr
    
    # Find outliers
    outliers = amounts[(amounts < lower_bound) | (amounts > upper_bound)]
    
    print(f"\nSubsidiary {name}:")
    print(f"  Normal range: €{lower_bound:,.0f} to €{upper_bound:,.0f}")
    print(f"  Number of outliers: {len(outliers)}")
    print(f"  Largest outlier: €{outliers.max():,.0f}")
    print(f"  Smallest outlier: €{outliers.min():,.0f}")

SALES OUTLIER ANALYSIS:

Subsidiary A:
  Normal range: €-6,964 to €11,719
  Number of outliers: 3855
  Largest outlier: €48,243,171
  Smallest outlier: €-7,111

Subsidiary G:
  Normal range: €-4,571 to €7,824
  Number of outliers: 17931
  Largest outlier: €5,677,990
  Smallest outlier: €-6,381

Subsidiary J:
  Normal range: €-7,360 to €13,195
  Number of outliers: 17224
  Largest outlier: €11,457,820
  Smallest outlier: €-1,059,676


In [13]:
# Purchase Outliers
print("\nPURCHASE OUTLIER ANALYSIS:")
for name, data in [('A', a_purchase), ('G', g_purchase), ('J', j_purchase)]:
    amounts = data['amount']
    amounts_nonzero = amounts[amounts > 0]  # Exclude zeros
    
    if len(amounts_nonzero) > 0:
        q1 = amounts_nonzero.quantile(0.25)
        q3 = amounts_nonzero.quantile(0.75)
        iqr = q3 - q1
        upper_bound = q3 + 1.5 * iqr
        
        outliers = amounts_nonzero[amounts_nonzero > upper_bound]
        
        print(f"\nSubsidiary {name}:")
        print(f"  Normal upper limit: €{upper_bound:,.0f}")
        print(f"  Purchase outliers: {len(outliers)}")
        print(f"  Largest purchase: €{outliers.max():,.0f}")


PURCHASE OUTLIER ANALYSIS:

Subsidiary A:
  Normal upper limit: €544,184
  Purchase outliers: 192
  Largest purchase: €131,516,729

Subsidiary G:
  Normal upper limit: €217,830
  Purchase outliers: 2297
  Largest purchase: €5,847,896

Subsidiary J:
  Normal upper limit: €643,182
  Purchase outliers: 348
  Largest purchase: €7,635,585


##### Sales Outliers:

Subsidiary A: 3,855 outliers out of 32,844 orders (12% are outliers!)
Subsidiary G: 17,931 outliers out of 135,546 orders (13% are outliers!)
Subsidiary J: 17,224 outliers out of 143,161 orders (12% are outliers!)

##### Extreme Sales:

Subsidiary A: €48.2 million sale (vs normal max €11,719!)
Subsidiary J: Massive refund of €1.06 million

##### Purchase Outliers:

Subsidiary A: €131.5 million purchase (vs normal max €544k!)
All subsidiaries have major purchase outliers

#### This data suggests we handle TWO types of business:

- Regular retail sales: €100-€10,000 (normal range)
- Large commercial projects: Multi-million euro solar installations

In [14]:
#Checking if the outliers are real:
# Let's investigate if these outliers are real business
# Look at the biggest sales for each subsidiary
print("TOP 5 LARGEST SALES:")

for name, data in [('A', a_sales), ('G', g_sales), ('J', j_sales)]:
    print(f"\nSubsidiary {name} - Top 5 sales:")
    top_sales = data.nlargest(5, 'amount')[['amount', 'qtyOrdered', 'tranDate']]
    print(top_sales)

# Look at some big purchases
print("\nTOP 3 LARGEST PURCHASES:")
for name, data in [('A', a_purchase), ('G', g_purchase), ('J', j_purchase)]:
    print(f"\nSubsidiary {name} - Top 3 purchases:")
    top_purchases = data.nlargest(3, 'amount')[['amount', 'qtyOrdered', 'tranDate']]
    print(top_purchases)

TOP 5 LARGEST SALES:

Subsidiary A - Top 5 sales:
           amount  qtyOrdered    tranDate
24460 48243171.48    48000.00  2023-09-14
31496 24179256.00      279.00  2023-11-28
16079 13847400.00      450.00  2023-06-05
12898  8138127.00     2922.00  2023-04-26
7945   4442641.56     3078.00  2023-02-06

Subsidiary G - Top 5 sales:
          amount  qtyOrdered    tranDate
80095 5677990.16    53352.00  2023-05-09
52541 5612378.54     1170.00  2022-12-12
40634 2557793.38     1188.00  2022-10-10
40660 2557793.38     1188.00  2022-10-10
43749 2557793.38     1188.00  2022-10-27

Subsidiary J - Top 5 sales:
         amount  qtyOrdered    tranDate
140 11457820.31     2997.00  2022-01-06
127  8014614.56     2997.00  2022-01-06
138  6570906.12     2997.00  2022-01-06
117  6108627.06     2997.00  2022-01-06
122  5514286.19     2997.00  2022-01-06

TOP 3 LARGEST PURCHASES:

Subsidiary A - Top 3 purchases:
          amount  qtyOrdered    tranDate
216 131516728.63     4608.00  2022-10-04
239 108779457

The above data shows that the outliers are real. 

- Realistic quantities: 48,000 units, 53,352 units (bulk solar equipment)
- Consistent dates: Real transaction dates in 2022-2023
- Business pattern: Large quantities = large amounts (makes sense)


Business Insight:
- Retail/Small Commercial: €100-€10k orders
- Large Solar Projects: €1M-€48M orders (solar farms, utility projects)

### 1. INITIAL FINANCIAL ANALYSIS

In [15]:
# Revenue analysis (including returns/refunds)
print("Revenue analysis (including returns):")
a_revenue = a_sales['amount'].sum()
g_revenue = g_sales['amount'].sum()
j_revenue = j_sales['amount'].sum()

print(f"Subsidiary A: €{a_revenue:,.0f}")
print(f"Subsidiary G: €{g_revenue:,.0f}")
print(f"Subsidiary J: €{j_revenue:,.0f}")

Revenue analysis (including returns):
Subsidiary A: €432,040,930
Subsidiary G: €1,342,551,182
Subsidiary J: €1,714,487,452


In [16]:
# Check the outliers
print("\nLargest sales transactions:")
print(f"A largest sale: €{a_sales['amount'].max():,.0f}")
print(f"G largest sale: €{g_sales['amount'].max():,.0f}")
print(f"J largest sale: €{j_sales['amount'].max():,.0f}")


Largest sales transactions:
A largest sale: €48,243,171
G largest sale: €5,677,990
J largest sale: €11,457,820


In [20]:
# Check the big refunds
print("\nLargest refunds (negative amounts):")
print(f"A largest refund: €{a_sales['amount'].min():,.0f}")
print(f"G largest refund: €{g_sales['amount'].min():,.0f}")
print(f"J largest refund: €{j_sales['amount'].min():,.0f}")


Largest refunds (negative amounts):
A largest refund: €-7,111
G largest refund: €-6,381
J largest refund: €-1,059,676


In [21]:
# Purchase totals
a_purchases = a_purchase['amount'].sum()
g_purchases = g_purchase['amount'].sum()
j_purchases = j_purchase['amount'].sum()

print(f"\nPurchase totals:")
print(f"Subsidiary A: €{a_purchases:,.0f}")
print(f"Subsidiary G: €{g_purchases:,.0f}")
print(f"Subsidiary J: €{j_purchases:,.0f}")


Purchase totals:
Subsidiary A: €680,530,916
Subsidiary G: €1,139,240,128
Subsidiary J: €1,412,641,618


In [22]:
# Quick profitability check
print(f"\nQuick profitability check (Revenue - Purchases):")
print(f"Subsidiary A: €{a_revenue - a_purchases:,.0f}")
print(f"Subsidiary G: €{g_revenue - g_purchases:,.0f}")
print(f"Subsidiary J: €{j_revenue - j_purchases:,.0f}")


Quick profitability check (Revenue - Purchases):
Subsidiary A: €-248,489,986
Subsidiary G: €203,311,054
Subsidiary J: €301,845,834


Subsidiary A shows NEGATIVE profitability!

### 2: INVESTIGATING THE ANOMALY

In [23]:
print("Checking purchase order statuses - maybe canceled orders are included?")

print("\nSubsidiary A Purchase Order Status:")
a_status_counts = a_purchase['status'].value_counts()
print(a_status_counts)

print("\nSubsidiary G Purchase Order Status:")
g_status_counts = g_purchase['status'].value_counts()
print(g_status_counts)

print("\nSubsidiary J Purchase Order Status:")
j_status_counts = j_purchase['status'].value_counts()
print(j_status_counts)

Checking purchase order statuses - maybe canceled orders are included?

Subsidiary A Purchase Order Status:
status
Fully Billed              1733
Canceled                   310
Approved by Supervisor      80
Approved                    48
Name: count, dtype: int64

Subsidiary G Purchase Order Status:
status
Fully Billed              12999
Closed                     1117
Rejected by Supervisor       74
Pending Bill                 34
Partially Received           19
Pending Receipt               1
Name: count, dtype: int64

Subsidiary J Purchase Order Status:
status
Fully Billed                   8654
Closed                          803
Rejected by Supervisor           45
Pending Supervisor Approval      10
Name: count, dtype: int64


In [25]:
# Calculate only COMPLETED purchases (not canceled/pending)
print("COMPLETED PURCHASES ONLY:")

# Filter for only 'Fully Billed' status (completed purchases)
a_completed_purchases = a_purchase[a_purchase['status'] == 'Fully Billed']['amount'].sum()
g_completed_purchases = g_purchase[g_purchase['status'] == 'Fully Billed']['amount'].sum()
j_completed_purchases = j_purchase[j_purchase['status'] == 'Fully Billed']['amount'].sum()

print(f"A completed purchases: €{a_completed_purchases:,.0f}")
print(f"G completed purchases: €{g_completed_purchases:,.0f}")
print(f"J completed purchases: €{j_completed_purchases:,.0f}")

COMPLETED PURCHASES ONLY:
A completed purchases: €293,397,946
G completed purchases: €997,520,297
J completed purchases: €1,163,441,535


In [26]:
# Recalculate CORRECT profitability
print("\nCORRECTED PROFITABILITY (using completed purchases only):")
a_correct_profit = a_revenue - a_completed_purchases
g_correct_profit = g_revenue - g_completed_purchases
j_correct_profit = j_revenue - j_completed_purchases

print(f"Subsidiary A: €{a_correct_profit:,.0f}")
print(f"Subsidiary G: €{g_correct_profit:,.0f}")
print(f"Subsidiary J: €{j_correct_profit:,.0f}")


CORRECTED PROFITABILITY (using completed purchases only):
Subsidiary A: €138,642,985
Subsidiary G: €345,030,885
Subsidiary J: €551,045,917


In [27]:
# Calculate profit margins
print("\nPROFIT MARGINS:")
print(f"Subsidiary A: {(a_correct_profit/a_revenue)*100:.1f}%")
print(f"Subsidiary G: {(g_correct_profit/g_revenue)*100:.1f}%")
print(f"Subsidiary J: {(j_correct_profit/j_revenue)*100:.1f}%")


PROFIT MARGINS:
Subsidiary A: 32.1%
Subsidiary G: 25.7%
Subsidiary J: 32.1%


### FINANCIAL PERFORMANCE ANALYSIS - KEY FINDINGS

#### Overall Performance Ranking:
1. **Subsidiary J**: €1.71B revenue, €551M profit (32.1% margin) - BEST PERFORMER
2. **Subsidiary G**: €1.34B revenue, €345M profit (25.7% margin) - SOLID PERFORMER  
3. **Subsidiary A**: €432M revenue, €139M profit (32.1% margin) - SMALLEST BUT EFFICIENT

#### Critical Data Discovery:
- **Initial analysis showed Subsidiary A with -€248M loss (WRONG)**
- **Root cause**: 310 canceled purchase orders (€387M) were incorrectly included
- **Corrected analysis**: All subsidiaries are profitable

### Operational Issues Identified:
- **Subsidiary A**: 14% purchase order cancellation rate (needs improvement)
- **Subsidiaries G & J**: <1% cancellation rate (efficient operations)

### Business Model Insights:
- **All subsidiaries operate retail + large commercial solar projects**
- **Large outlier transactions (€1M-€48M) are legitimate solar farm installations**
- **Subsidiaries A & J**: Higher profit margins (32.1%)
- **Subsidiary G**: Lower margin (25.7%) but largest scale

### Executive Recommendations:
1. **Invest in Subsidiary J expansion** (best overall performance)
2. **Improve Subsidiary A procurement processes** (reduce cancellations)
3. **Leverage Subsidiary A's efficiency model** for G & J
4. **Focus on large commercial solar projects** (higher value transactions)

### 3: DEEPER DATA QUALITY INVESTIGATION

In [29]:
print("If purchase orders had canceled items, maybe sales orders do too...")

print("\nSubsidiary A Sales Order Status:")
a_sales_status = a_sales['status'].value_counts()
print(a_sales_status)

print("\nSubsidiary G Sales Order Status:")
g_sales_status = g_sales['status'].value_counts()
print(g_sales_status)

print("\nSubsidiary J Sales Order Status:")
j_sales_status = j_sales['status'].value_counts()
print(j_sales_status)

If purchase orders had canceled items, maybe sales orders do too...

Subsidiary A Sales Order Status:
status
Billed                         26059
Cancelled                       6233
Pending Billing                  381
Created                          150
Approved by Supervisor            16
Goods Reserved                     3
Pending Supervisor Approval        2
Name: count, dtype: int64

Subsidiary G Sales Order Status:
status
Billed       126595
Closed         8068
Cancelled       883
Name: count, dtype: int64

Subsidiary J Sales Order Status:
status
Billed       123415
Closed        17222
Cancelled      2524
Name: count, dtype: int64


In [34]:
# Calculate the actual cancellation rates
a_cancel_count = (a_sales['status'] == 'Cancelled').sum()
g_cancel_count = (g_sales['status'] == 'Cancelled').sum()
j_cancel_count = (j_sales['status'] == 'Cancelled').sum()

a_cancel_rate = (a_cancel_count / len(a_sales)) * 100
g_cancel_rate = (g_cancel_count / len(g_sales)) * 100
j_cancel_rate = (j_cancel_count / len(j_sales)) * 100

print(f"\nSubsidiary A has {a_cancel_count:,} canceled sales orders!")
print("Sales Order Issues Found:")
print(f"Subsidiary A: {a_cancel_count:,} canceled sales orders ({a_cancel_rate:.1f}% cancellation rate!)")
print(f"Subsidiary G: {g_cancel_count:,} canceled sales orders ({g_cancel_rate:.1f}% cancellation rate)")
print(f"Subsidiary J: {j_cancel_count:,} canceled sales orders ({j_cancel_rate:.1f}% cancellation rate)")


Subsidiary A has 6,233 canceled sales orders!
Sales Order Issues Found:
Subsidiary A: 6,233 canceled sales orders (19.0% cancellation rate!)
Subsidiary G: 883 canceled sales orders (0.7% cancellation rate)
Subsidiary J: 2,524 canceled sales orders (1.8% cancellation rate)


In [35]:
# Show completion rates for context
a_completion_rate = ((a_sales['status'] == 'Billed').sum() / len(a_sales)) * 100
g_completion_rate = ((g_sales['status'].isin(['Billed', 'Closed'])).sum() / len(g_sales)) * 100
j_completion_rate = ((j_sales['status'].isin(['Billed', 'Closed'])).sum() / len(j_sales)) * 100

print(f"\nSALES COMPLETION RATES:")
print(f"Subsidiary A: {a_completion_rate:.1f}% completion rate")
print(f"Subsidiary G: {g_completion_rate:.1f}% completion rate") 
print(f"Subsidiary J: {j_completion_rate:.1f}% completion rate")

print(f"\nSubsidiary A has serious operational issues:")
print(f"   • Only {a_completion_rate:.1f}% of sales orders complete successfully")
print(f"   • {a_cancel_rate:.1f}% cancellation rate is extremely high")
print(f"   • This suggests major problems with order fulfillment processes")


SALES COMPLETION RATES:
Subsidiary A: 79.3% completion rate
Subsidiary G: 99.3% completion rate
Subsidiary J: 98.2% completion rate

Subsidiary A has serious operational issues:
   • Only 79.3% of sales orders complete successfully
   • 19.0% cancellation rate is extremely high
   • This suggests major problems with order fulfillment processes


Now that we have identified data quality issues in both sales AND purchases, will recalculate everything using only COMPLETED transactions.

In [36]:
# Cleaning sales data - only completed transactions
# Filter to only completed sales (Billed/Closed status)
a_sales_clean = a_sales[a_sales['status'].isin(['Billed'])]
g_sales_clean = g_sales[g_sales['status'].isin(['Billed', 'Closed'])]
j_sales_clean = j_sales[j_sales['status'].isin(['Billed', 'Closed'])]

print("\nCleaned sales order counts:")
print(f"Subsidiary A: {len(a_sales_clean):,} completed (vs {len(a_sales):,} total)")
print(f"Subsidiary G: {len(g_sales_clean):,} completed (vs {len(g_sales):,} total)")
print(f"Subsidiary J: {len(j_sales_clean):,} completed (vs {len(j_sales):,} total)")


Cleaned sales order counts:
Subsidiary A: 26,059 completed (vs 32,844 total)
Subsidiary G: 134,663 completed (vs 135,546 total)
Subsidiary J: 140,637 completed (vs 143,161 total)


In [37]:
# Recalculate CORRECTED revenue with clean data
print("\nCORRECTED REVENUE (completed sales only):")
a_revenue_clean = a_sales_clean['amount'].sum()
g_revenue_clean = g_sales_clean['amount'].sum()
j_revenue_clean = j_sales_clean['amount'].sum()

print(f"Subsidiary A: €{a_revenue_clean:,.0f}")
print(f"Subsidiary G: €{g_revenue_clean:,.0f}")
print(f"Subsidiary J: €{j_revenue_clean:,.0f}")


CORRECTED REVENUE (completed sales only):
Subsidiary A: €227,891,447
Subsidiary G: €1,311,301,519
Subsidiary J: €1,550,822,997


In [38]:
# Recalculate completed purchases (from our earlier analysis)
print("\nRECALCULATING COMPLETED PURCHASES:")

a_completed_purchases = a_purchase[a_purchase['status'] == 'Fully Billed']['amount'].sum()
g_completed_purchases = g_purchase[g_purchase['status'] == 'Fully Billed']['amount'].sum()
j_completed_purchases = j_purchase[j_purchase['status'] == 'Fully Billed']['amount'].sum()

print(f"A completed purchases: €{a_completed_purchases:,.0f}")
print(f"G completed purchases: €{g_completed_purchases:,.0f}")
print(f"J completed purchases: €{j_completed_purchases:,.0f}")


RECALCULATING COMPLETED PURCHASES:
A completed purchases: €293,397,946
G completed purchases: €997,520,297
J completed purchases: €1,163,441,535


In [39]:
print("\nFINAL CORRECTED PROFITABILITY:")
a_final_profit = a_revenue_clean - a_completed_purchases
g_final_profit = g_revenue_clean - g_completed_purchases
j_final_profit = j_revenue_clean - j_completed_purchases

print(f"Subsidiary A: €{a_final_profit:,.0f}")
print(f"Subsidiary G: €{g_final_profit:,.0f}")
print(f"Subsidiary J: €{j_final_profit:,.0f}")


FINAL CORRECTED PROFITABILITY:
Subsidiary A: €-65,506,498
Subsidiary G: €313,781,222
Subsidiary J: €387,381,461


In [40]:
print("\nFINAL PROFIT MARGINS:")
a_margin = (a_final_profit / a_revenue_clean) * 100
g_margin = (g_final_profit / g_revenue_clean) * 100
j_margin = (j_final_profit / j_revenue_clean) * 100

print(f"Subsidiary A: {a_margin:.1f}%")
print(f"Subsidiary G: {g_margin:.1f}%")
print(f"Subsidiary J: {j_margin:.1f}%")


FINAL PROFIT MARGINS:
Subsidiary A: -28.7%
Subsidiary G: 23.9%
Subsidiary J: 25.0%


In [43]:
print("\nSubsidiary A is actually LOSING MONEY!")
print("The complete investigation revealed:")
print(f"• Actual revenue: €{a_revenue_clean:,.0f} (much lower than initial €{a_revenue:,.0f})")
print(f"• Net loss: €{abs(a_final_profit):,.0f}")
print("• 19% sales cancellation rate + 14% purchase cancellation rate")
print("• This indicates serious operational and management problems")


Subsidiary A is actually LOSING MONEY!
The complete investigation revealed:
• Actual revenue: €227,891,447 (much lower than initial €432,040,930)
• Net loss: €65,506,498
• 19% sales cancellation rate + 14% purchase cancellation rate
• This indicates serious operational and management problems


### 4. BUSINESS SEGMENT ANALYSIS

In [44]:
# Define thresholds for retail vs large projects (based on our earlier outlier analysis)
thresholds = {'A': 11719, 'G': 7824, 'J': 13195}

for name, sales_data in [('A', a_sales_clean), ('G', g_sales_clean), ('J', j_sales_clean)]:
    
    threshold = thresholds[name]
    
    # Separate retail from large projects
    retail_sales = sales_data[sales_data['amount'] <= threshold]
    large_projects = sales_data[sales_data['amount'] > threshold]
    
    total_revenue = sales_data['amount'].sum()
    retail_revenue = retail_sales['amount'].sum()
    project_revenue = large_projects['amount'].sum()
    
    print(f"\nSubsidiary {name} Business Breakdown:")
    print(f"  Total Revenue: €{total_revenue:,.0f}")
    print(f"  Retail Business: €{retail_revenue:,.0f} ({len(retail_sales):,} orders)")
    print(f"  Large Projects: €{project_revenue:,.0f} ({len(large_projects):,} projects)")
    
    if len(large_projects) > 0:
        project_percentage = (project_revenue / total_revenue) * 100
        avg_project_size = project_revenue / len(large_projects)
        print(f"  % from Large Projects: {project_percentage:.1f}%")
        print(f"  Average Project Size: €{avg_project_size:,.0f}")
        
        # Show some example large projects
        if len(large_projects) > 0:
            largest_projects = large_projects.nlargest(3, 'amount')
            print(f"  Top 3 project sizes: €{largest_projects['amount'].iloc[0]:,.0f}, €{largest_projects['amount'].iloc[1]:,.0f}, €{largest_projects['amount'].iloc[2]:,.0f}")
    else:
        print(f"  % from Large Projects: 0%")


Subsidiary A Business Breakdown:
  Total Revenue: €227,891,447
  Retail Business: €42,894,896 (23,194 orders)
  Large Projects: €184,996,551 (2,865 projects)
  % from Large Projects: 81.2%
  Average Project Size: €64,571
  Top 3 project sizes: €4,442,642, €3,126,303, €2,707,308

Subsidiary G Business Breakdown:
  Total Revenue: €1,311,301,519
  Retail Business: €136,903,972 (117,013 orders)
  Large Projects: €1,174,397,547 (17,650 projects)
  % from Large Projects: 89.6%
  Average Project Size: €66,538
  Top 3 project sizes: €5,677,990, €2,557,793, €2,557,793

Subsidiary J Business Breakdown:
  Total Revenue: €1,550,822,997
  Retail Business: €290,372,253 (124,087 orders)
  Large Projects: €1,260,450,744 (16,550 projects)
  % from Large Projects: 81.3%
  Average Project Size: €76,160
  Top 3 project sizes: €4,753,114, €3,050,689, €2,520,478


All Subsidiaries are **Project-Focused:**

- Subsidiary G: 89.6% revenue from large projects (most project-focused)
- Subsidiary A: 81.2% revenue from large projects
- Subsidiary J: 81.3% revenue from large projects

**Project Economics:**

- Subsidiary J: Highest average project size (€76,160)
- Subsidiary G: €66,538 average project
- Subsidiary A: €64,571 average project

This explains why Subsidiary A is losing money - they're in the high-value commercial solar business but have terrible operational efficiency

### 5. Product Category Analysis:

In [46]:
# Show product categories for each subsidiary
print("\nProduct categories available:")

print("\nSubsidiary A categories:")
a_categories = a_items['itemCategory'].value_counts()
print(a_categories)

print("\nSubsidiary G categories:")
g_categories = g_items['itemCategory'].value_counts()
print(g_categories)

print("\nSubsidiary J categories:")
j_categories = j_items['itemCategory'].value_counts()
print(j_categories)


Product categories available:

Subsidiary A categories:
itemCategory
Racking                 179
Inverters               174
Other                   131
Storage                  53
Inverter Accessories     36
Charging Stations        34
Modules                  29
Name: count, dtype: int64

Subsidiary G categories:
itemCategory
Other                   478
Inverters               342
Racking                 338
Inverter Accessories    232
Modules                 210
Charging Stations        71
Storage                  49
Shelter                  17
Name: count, dtype: int64

Subsidiary J categories:
itemCategory
Racking                 562
Inverters               339
Modules                 231
Inverter Accessories    210
Other                   173
Charging Stations       127
Storage                  65
Name: count, dtype: int64


In [47]:
# Calculate revenue by product category
print("\nREVENUE BY PRODUCT CATEGORY:")

for name, sales_data, items_data in [('A', a_sales_clean, a_items), 
                                     ('G', g_sales_clean, g_items), 
                                     ('J', j_sales_clean, j_items)]:
    
    print(f"\nSubsidiary {name} - Revenue by Product Category:")
    
    # Creating lookup dictionary for item categories
    item_category_lookup = dict(zip(items_data['guid'], items_data['itemCategory']))
    
    # Adding categories to sales data
    sales_with_categories = sales_data.copy()
    sales_with_categories['category'] = sales_with_categories['itemGuid'].map(item_category_lookup)
    
    # Calculating revenue by category
    category_revenue = sales_with_categories.groupby('category')['amount'].sum().sort_values(ascending=False)
    
    #All categories with percentages
    total_revenue = sales_with_categories['amount'].sum()
    for category, revenue in category_revenue.items():
        if pd.notna(category):  # Skip NaN categories
            percentage = (revenue / total_revenue) * 100
            order_count = sales_with_categories[sales_with_categories['category'] == category].shape[0]
            avg_order = revenue / order_count if order_count > 0 else 0
            print(f"  {category}: €{revenue:,.0f} ({percentage:.1f}%) - {order_count:,} orders, €{avg_order:.0f} avg")



REVENUE BY PRODUCT CATEGORY:

Subsidiary A - Revenue by Product Category:
  Storage: €127,859,749 (56.1%) - 4,079 orders, €31346 avg
  Inverters: €57,024,335 (25.0%) - 4,339 orders, €13142 avg
  Racking: €16,218,302 (7.1%) - 7,862 orders, €2063 avg
  Modules: €15,419,313 (6.8%) - 851 orders, €18119 avg
  Other: €7,079,986 (3.1%) - 7,860 orders, €901 avg
  Inverter Accessories: €3,542,982 (1.6%) - 807 orders, €4390 avg
  Charging Stations: €746,781 (0.3%) - 261 orders, €2861 avg

Subsidiary G - Revenue by Product Category:
  Modules: €813,090,227 (62.0%) - 10,908 orders, €74541 avg
  Inverters: €232,395,498 (17.7%) - 14,466 orders, €16065 avg
  Racking: €102,148,471 (7.8%) - 69,149 orders, €1477 avg
  Storage: €68,009,708 (5.2%) - 1,806 orders, €37658 avg
  Inverter Accessories: €45,962,616 (3.5%) - 19,060 orders, €2411 avg
  Other: €41,880,875 (3.2%) - 17,995 orders, €2327 avg
  Shelter: €5,286,199 (0.4%) - 831 orders, €6361 avg
  Charging Stations: €2,527,926 (0.2%) - 448 orders, €56

**BUSINESS QUESTION: Which solar products drive revenue for each subsidiary?**

This analysis reveals each subsidiary's product specialization and explains their performance differences:

**KEY FINDINGS:**
- Subsidiary A: Storage specialist (56% revenue, €31k avg order)
  - Premium battery storage focus but operational failures hurt profitability
  
- Subsidiary G: Module specialist (62% revenue, €74k avg order) 
  - Highest-value solar panel installations with excellent operational efficiency
  
- Subsidiary J: Diversified portfolio (39% modules, 36% inverters, 15% storage)
  - Balanced approach reduces risk and enables sustainable growth

**STRATEGIC IMPLICATIONS:**
- A has the right premium strategy but needs operational fixes
- G found their profitable niche and should be expanded
- J's diversification model provides stability and growth potential

This product mix analysis directly informs:
- Investment allocation decisions
- Operational improvement priorities  
- Cross-subsidiary knowledge sharing opportunities

### 5.1. OPERATIONAL EFFICIENCY METRICS

**BUSINESS QUESTION: How efficiently do subsidiaries manage inventory and operations?**

This analysis examines:
- Inventory turnover rates (working capital efficiency)
- Stock management effectiveness  
- Operational capacity utilization
- Supply chain performance indicators

These metrics help operations managers optimize processes and help executives make informed investment decisions about working capital and operational improvements.

In [48]:
#Analyzing inventory management and operational efficiency
# Inventory Analysis
print("\nINVENTORY MANAGEMENT EFFICIENCY:")

for name, inventory_data, revenue in [('A', a_inventory, a_revenue_clean), 
                                      ('G', g_inventory, g_revenue_clean), 
                                      ('J', j_inventory, j_revenue_clean)]:
    
    print(f"\nSubsidiary {name} Inventory Metrics:")
    
    # Basic inventory stats
    total_on_hand = inventory_data['quantityOnHand'].sum()
    total_available = inventory_data['quantityAvailable'].sum()
    total_committed = inventory_data['quantityCommitted'].sum()
    
    # Calculate inventory value
    inventory_value = (inventory_data['quantityOnHand'] * inventory_data['averageCost']).sum()
    
    print(f"  Total items on hand: {total_on_hand:,}")
    print(f"  Total available for sale: {total_available:,}")
    print(f"  Total committed to orders: {total_committed:,}")
    print(f"  Total inventory value: €{inventory_value:,.0f}")
    
    # Calculate inventory turnover
    if inventory_value > 0:
        turnover_ratio = revenue / inventory_value
        days_of_inventory = 365 / turnover_ratio if turnover_ratio > 0 else 0
        print(f"  Inventory turnover: {turnover_ratio:.1f}x per year")
        print(f"  Days of inventory on hand: {days_of_inventory:.0f} days")
        
        # Efficiency assessment
        if turnover_ratio > 10:
            efficiency = "EXCELLENT"
        elif turnover_ratio > 5:
            efficiency = "GOOD" 
        elif turnover_ratio > 2:
            efficiency = "AVERAGE"
        else:
            efficiency = "POOR - NEEDS IMPROVEMENT"
        
        print(f"  Inventory efficiency: {efficiency}")
    
    # Stock availability ratio
    if total_on_hand > 0:
        availability_ratio = (total_available / total_on_hand) * 100
        print(f"  Stock availability: {availability_ratio:.1f}%")


INVENTORY MANAGEMENT EFFICIENCY:

Subsidiary A Inventory Metrics:
  Total items on hand: 750,673
  Total available for sale: 683,499
  Total committed to orders: 67,174
  Total inventory value: €37,295,940
  Inventory turnover: 6.1x per year
  Days of inventory on hand: 60 days
  Inventory efficiency: GOOD
  Stock availability: 91.1%

Subsidiary G Inventory Metrics:
  Total items on hand: 4,683,802
  Total available for sale: 3,867,185
  Total committed to orders: 816,631
  Total inventory value: €118,394,875
  Inventory turnover: 11.1x per year
  Days of inventory on hand: 33 days
  Inventory efficiency: EXCELLENT
  Stock availability: 82.6%

Subsidiary J Inventory Metrics:
  Total items on hand: 3,831,274
  Total available for sale: 3,180,545
  Total committed to orders: 650,743
  Total inventory value: €118,365,986
  Inventory turnover: 13.1x per year
  Days of inventory on hand: 28 days
  Inventory efficiency: EXCELLENT
  Stock availability: 83.0%


In [49]:
# Order Processing Efficiency
print("\nORDER PROCESSING EFFICIENCY:")

for name, sales_clean, sales_total in [('A', a_sales_clean, a_sales), 
                                       ('G', g_sales_clean, g_sales), 
                                       ('J', j_sales_clean, j_sales)]:
    
    print(f"\nSubsidiary {name} Processing Metrics:")
    
    # Order completion efficiency
    completion_rate = (len(sales_clean) / len(sales_total)) * 100
    
    # Average order processing (using order counts as proxy)
    avg_orders_per_day = len(sales_clean) / 365  # Assuming 1 year of data
    
    print(f"  Order completion rate: {completion_rate:.1f}%")
    print(f"  Average daily orders processed: {avg_orders_per_day:.0f}")
    
    # Efficiency rating
    if completion_rate > 95:
        rating = "EXCELLENT"
    elif completion_rate > 85:
        rating = "GOOD"
    elif completion_rate > 75:
        rating = "AVERAGE"
    else:
        rating = "POOR - CRITICAL ISSUE"
    
    print(f"  Processing efficiency: {rating}")



ORDER PROCESSING EFFICIENCY:

Subsidiary A Processing Metrics:
  Order completion rate: 79.3%
  Average daily orders processed: 71
  Processing efficiency: AVERAGE

Subsidiary G Processing Metrics:
  Order completion rate: 99.3%
  Average daily orders processed: 369
  Processing efficiency: EXCELLENT

Subsidiary J Processing Metrics:
  Order completion rate: 98.2%
  Average daily orders processed: 385
  Processing efficiency: EXCELLENT


#### Operational Discoveries:
**Subsidiary G & J: World-Class Operations**

- G: 11.1x inventory turnover, 33 days inventory, 99.3% order completion
- J: 13.1x inventory turnover, 28 days inventory, 98.2% order completion
Both: EXCELLENT operational efficiency across all metrics

**Subsidiary A: Operational Crisis Confirmed**

- Inventory: Only 6.1x turnover (60 days vs 28-33 for others)
- Orders: Only 79.3% completion rate (vs 98-99% for others)
- Scale: 71 orders/day vs 369-385 for others

#### Key Business Insights:
1. Cash Flow Impact:
- A ties up cash for 60 days vs 28-33 days (inefficient)
- G & J convert inventory to cash twice as fast

2. Operational Excellence:
- G & J are benchmark performers (99%+ completion)
- A has systemic process failures (20% order failure rate)

3. Scale Efficiency:
- G & J process 5x more daily orders than A
- Higher volume = better operational leverage

### 6. EXECUTIVE SUMMARY AND STRATEGIC RECOMMENDATIONS

In [51]:
print("FINAL BUSINESS PERFORMANCE RANKING:")

# Create performance summary
subsidiaries_performance = [
    ('J', j_revenue_clean, j_final_profit, j_margin, 'BEST PERFORMER'),
    ('G', g_revenue_clean, g_final_profit, g_margin, 'SOLID PERFORMER'),
    ('A', a_revenue_clean, a_final_profit, a_margin, 'CRISIS - IMMEDIATE ACTION NEEDED')
]

for i, (name, revenue, profit, margin, status) in enumerate(subsidiaries_performance, 1):
    print(f"{i}. Subsidiary {name}: €{revenue:,.0f} revenue, €{profit:,.0f} {'profit' if profit > 0 else 'LOSS'} ({margin:.1f}% margin)")
    print(f"   Status: {status}")

FINAL BUSINESS PERFORMANCE RANKING:
1. Subsidiary J: €1,550,822,997 revenue, €387,381,461 profit (25.0% margin)
   Status: BEST PERFORMER
2. Subsidiary G: €1,311,301,519 revenue, €313,781,222 profit (23.9% margin)
   Status: SOLID PERFORMER
3. Subsidiary A: €227,891,447 revenue, €-65,506,498 LOSS (-28.7% margin)
   Status: CRISIS - IMMEDIATE ACTION NEEDED


In [53]:
print(f"\nOPERATIONAL EFFICIENCY COMPARISON:")
print(f"• Subsidiary A: {a_cancel_rate:.1f}% sales cancellation, 6.1x inventory turnover - CRITICAL PROBLEMS")
print(f"• Subsidiary G: {g_cancel_rate:.1f}% sales cancellation, 11.1x inventory turnover - EXCELLENT")
print(f"• Subsidiary J: {j_cancel_rate:.1f}% sales cancellation, 13.1x inventory turnover - EXCELLENT")


OPERATIONAL EFFICIENCY COMPARISON:
• Subsidiary A: 19.0% sales cancellation, 6.1x inventory turnover - CRITICAL PROBLEMS
• Subsidiary G: 0.7% sales cancellation, 11.1x inventory turnover - EXCELLENT
• Subsidiary J: 1.8% sales cancellation, 13.1x inventory turnover - EXCELLENT


**BUSINESS MODEL INSIGHTS**
- All subsidiaries focus on large commercial solar projects (80%+ revenue)
- Subsidiary G: Module specialist (62% revenue) - excellent operational efficiency 
- Subsidiary J: Diversified portfolio (39% modules, 36% inverters) - balanced & efficient

**STRATEGIC RECOMMENDATIONS**
URGENT ACTIONS (Subsidiary A):
- Fix operational crisis: 20% order failure rate unacceptable
- Improve inventory management: 2x slower than benchmarks
- Consider management changes or process overhaul
- Focus on order fulfillment before growth initiatives

**GROWTH OPPORTUNITIES:**
- Expand Subsidiary J operations (best overall performance)
- Leverage G's module expertise and operational excellence
- Focus investment on large commercial solar projects
- Apply G & J operational practices company-wide

**OPERATIONAL IMPROVEMENTS:**
- Benchmark A against G & J processes
- Standardize inventory management across subsidiaries
- Implement order fulfillment tracking and alerts
- Improve supplier relationship management

In [56]:
print(f"\nFINANCIAL IMPACT ANALYSIS:")
total_profit = j_final_profit + g_final_profit + a_final_profit
potential_improvement = abs(a_final_profit) * 1.5  # Conservative estimate if A fixed

print(f"• Current total company profit: €{total_profit:,.0f}")
print(f"• A's losses offset {abs(a_final_profit)/(g_final_profit + j_final_profit)*100:.0f}% of G&J's profits")
print(f"• Potential company profit if A fixed: €{total_profit + potential_improvement:,.0f}+")
print(f"• Working capital savings from A efficiency: €15-20M+ annually")


FINANCIAL IMPACT ANALYSIS:
• Current total company profit: €635,656,186
• A's losses offset 9% of G&J's profits
• Potential company profit if A fixed: €733,915,933+
• Working capital savings from A efficiency: €15-20M+ annually


#### IMMEDIATE ACTIONS (NEXT 30 DAYS)
- Conduct deep-dive operational audit at Subsidiary A
- Document G & J best practices for replication
- Implement weekly operational metrics reporting
- Establish improvement targets and timeline for A

In [ ]:
**Perfect!** Now let’s move to **Tableau Dashboard Creation**. First, we need to prepare clean data exports for Tableau.

## **Step 1: Export Data for Tableau**

```python
# =============================================================================
# SECTION 8: PREPARE DATA FOR TABLEAU DASHBOARD
# =============================================================================

print("📊 PREPARING DATA EXPORTS FOR TABLEAU")
print("=" * 50)

# 1. Executive Summary Table
executive_summary = pd.DataFrame({
    'Subsidiary': ['A', 'G', 'J'],
    'Revenue': [a_revenue_clean, g_revenue_clean, j_revenue_clean],
    'Profit': [a_final_profit, g_final_profit, j_final_profit],
    'Profit_Margin': [a_margin, g_margin, j_margin],
    'Sales_Orders': [len(a_sales_clean), len(g_sales_clean), len(j_sales_clean)],
    'Cancellation_Rate': [a_cancel_rate, g_cancel_rate, j_cancel_rate],
    'Inventory_Turnover': [6.1, 11.1, 13.1],
    'Days_Inventory': [60, 33, 28],
    'Status': ['Crisis', 'Good', 'Best']
})

# 2. Business Segment Data
business_segments = pd.DataFrame({
    'Subsidiary': ['A', 'A', 'G', 'G', 'J', 'J'],
    'Segment': ['Retail', 'Large Projects', 'Retail', 'Large Projects', 'Retail', 'Large Projects'],
    'Revenue': [42894896, 184996551, 136903972, 1174397547, 290372253, 1260450744],
    'Order_Count': [23194, 2865, 117013, 17650, 124087, 16550],
    'Avg_Order_Size': [1849, 64571, 1170, 66538, 2340, 76160]
})

# 3. Product Category Data
product_categories = pd.DataFrame({
    'Subsidiary': ['A', 'A', 'A', 'A', 'A', 'A', 'A',
                   'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G',
                   'J', 'J', 'J', 'J', 'J', 'J', 'J'],
    'Category': ['Storage', 'Inverters', 'Racking', 'Modules', 'Other', 'Inverter Accessories', 'Charging Stations',
                 'Modules', 'Inverters', 'Racking', 'Storage', 'Inverter Accessories', 'Other', 'Shelter', 'Charging Stations',
                 'Modules', 'Inverters', 'Storage', 'Inverter Accessories', 'Racking', 'Other', 'Charging Stations'],
    'Revenue': [127859749, 57024335, 16218302, 15419313, 7079986, 3542982, 746781,
                813090227, 232395498, 102148471, 68009708, 45962616, 41880875, 5286199, 2527926,
                597882364, 557685767, 237023724, 85944713, 49763777, 14686747, 7835905],
    'Revenue_Percentage': [56.1, 25.0, 7.1, 6.8, 3.1, 1.6, 0.3,
                          62.0, 17.7, 7.8, 5.2, 3.5, 3.2, 0.4, 0.2,
                          38.6, 36.0, 15.3, 5.5, 3.2, 0.9, 0.5]
})

# Export all datasets
executive_summary.to_csv('executive_summary.csv', index=False)
business_segments.to_csv('business_segments.csv', index=False)
product_categories.to_csv('product_categories.csv', index=False)

print("✅ Data exported successfully!")
print("Files created:")
print("  • executive_summary.csv - Main KPIs for C-Level dashboard")
print("  • business_segments.csv - Retail vs Projects analysis")
print("  • product_categories.csv - Product performance data")

print(f"\n📋 TABLEAU DASHBOARD SPECIFICATIONS:")
print("=" * 50)
print("DASHBOARD 1: C-LEVEL EXECUTIVE SUMMARY")
print("• KPI Cards: Revenue, Profit, Margin for each subsidiary")
print("• Performance Ranking: Visual ranking with status indicators") 
print("• Trend Charts: Revenue and profit trends (if time data available)")
print("• Key Metrics: Cancellation rates, inventory turnover")

print("\nDASHBOARD 2: OPERATIONS MANAGEMENT")
print("• Operational Efficiency: Cancellation rates, completion rates")
print("• Inventory Metrics: Turnover ratios, days of inventory")
print("• Product Performance: Revenue by category, order volumes")
print("• Process Monitoring: Order fulfillment tracking")

print("\nDASHBOARD 3: BUSINESS SEGMENT ANALYSIS") 
print("• Segment Breakdown: Retail vs Large Projects revenue split")
print("• Project Analytics: Average project size, project count")
print("• Product Categories: Revenue contribution by product type")
print("• Comparative Analysis: Cross-subsidiary comparisons")

print(f"\n🎯 NEXT STEPS:")
print("1. Open Tableau Public/Desktop")
print("2. Import the 3 CSV files created above")
print("3. Create the 3 dashboards as specified")
print("4. Publish to Tableau Public and get shareable link")
print("5. Write final summary report")
```

**Run this to create the data files for Tableau!**

**After this runs successfully, we’ll move to Tableau and I’ll guide you through creating the dashboards step by step.****Perfect!** Now let’s move to **Tableau Dashboard Creation**. First, we need to prepare clean data exports for Tableau.

## **Step 1: Export Data for Tableau**

```python
# =============================================================================
# SECTION 8: PREPARE DATA FOR TABLEAU DASHBOARD
# =============================================================================

print("📊 PREPARING DATA EXPORTS FOR TABLEAU")
print("=" * 50)

# 1. Executive Summary Table
executive_summary = pd.DataFrame({
    'Subsidiary': ['A', 'G', 'J'],
    'Revenue': [a_revenue_clean, g_revenue_clean, j_revenue_clean],
    'Profit': [a_final_profit, g_final_profit, j_final_profit],
    'Profit_Margin': [a_margin, g_margin, j_margin],
    'Sales_Orders': [len(a_sales_clean), len(g_sales_clean), len(j_sales_clean)],
    'Cancellation_Rate': [a_cancel_rate, g_cancel_rate, j_cancel_rate],
    'Inventory_Turnover': [6.1, 11.1, 13.1],
    'Days_Inventory': [60, 33, 28],
    'Status': ['Crisis', 'Good', 'Best']
})

# 2. Business Segment Data
business_segments = pd.DataFrame({
    'Subsidiary': ['A', 'A', 'G', 'G', 'J', 'J'],
    'Segment': ['Retail', 'Large Projects', 'Retail', 'Large Projects', 'Retail', 'Large Projects'],
    'Revenue': [42894896, 184996551, 136903972, 1174397547, 290372253, 1260450744],
    'Order_Count': [23194, 2865, 117013, 17650, 124087, 16550],
    'Avg_Order_Size': [1849, 64571, 1170, 66538, 2340, 76160]
})

# 3. Product Category Data
product_categories = pd.DataFrame({
    'Subsidiary': ['A', 'A', 'A', 'A', 'A', 'A', 'A',
                   'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G',
                   'J', 'J', 'J', 'J', 'J', 'J', 'J'],
    'Category': ['Storage', 'Inverters', 'Racking', 'Modules', 'Other', 'Inverter Accessories', 'Charging Stations',
                 'Modules', 'Inverters', 'Racking', 'Storage', 'Inverter Accessories', 'Other', 'Shelter', 'Charging Stations',
                 'Modules', 'Inverters', 'Storage', 'Inverter Accessories', 'Racking', 'Other', 'Charging Stations'],
    'Revenue': [127859749, 57024335, 16218302, 15419313, 7079986, 3542982, 746781,
                813090227, 232395498, 102148471, 68009708, 45962616, 41880875, 5286199, 2527926,
                597882364, 557685767, 237023724, 85944713, 49763777, 14686747, 7835905],
    'Revenue_Percentage': [56.1, 25.0, 7.1, 6.8, 3.1, 1.6, 0.3,
                          62.0, 17.7, 7.8, 5.2, 3.5, 3.2, 0.4, 0.2,
                          38.6, 36.0, 15.3, 5.5, 3.2, 0.9, 0.5]
})

# Export all datasets
executive_summary.to_csv('executive_summary.csv', index=False)
business_segments.to_csv('business_segments.csv', index=False)
product_categories.to_csv('product_categories.csv', index=False)

print("✅ Data exported successfully!")
print("Files created:")
print("  • executive_summary.csv - Main KPIs for C-Level dashboard")
print("  • business_segments.csv - Retail vs Projects analysis")
print("  • product_categories.csv - Product performance data")

print(f"\n📋 TABLEAU DASHBOARD SPECIFICATIONS:")
print("=" * 50)
print("DASHBOARD 1: C-LEVEL EXECUTIVE SUMMARY")
print("• KPI Cards: Revenue, Profit, Margin for each subsidiary")
print("• Performance Ranking: Visual ranking with status indicators") 
print("• Trend Charts: Revenue and profit trends (if time data available)")
print("• Key Metrics: Cancellation rates, inventory turnover")

print("\nDASHBOARD 2: OPERATIONS MANAGEMENT")
print("• Operational Efficiency: Cancellation rates, completion rates")
print("• Inventory Metrics: Turnover ratios, days of inventory")
print("• Product Performance: Revenue by category, order volumes")
print("• Process Monitoring: Order fulfillment tracking")

print("\nDASHBOARD 3: BUSINESS SEGMENT ANALYSIS") 
print("• Segment Breakdown: Retail vs Large Projects revenue split")
print("• Project Analytics: Average project size, project count")
print("• Product Categories: Revenue contribution by product type")
print("• Comparative Analysis: Cross-subsidiary comparisons")

print(f"\n🎯 NEXT STEPS:")
print("1. Open Tableau Public/Desktop")
print("2. Import the 3 CSV files created above")
print("3. Create the 3 dashboards as specified")
print("4. Publish to Tableau Public and get shareable link")
print("5. Write final summary report")
```

**Run this to create the data files for Tableau!**

**After this runs successfully, we’ll move to Tableau and I’ll guide you through creating the dashboards step by step.**